core/dbt/contracts/graph/nodes.py

```diff
@@ -492,6 +502,46 @@ def depends_on_nodes(self):
     def depends_on_macros(self):
         return self.depends_on.macros
 
+    def build_contract_checksum(self):
+        # We don't need to construct the checksum if the model does not
+        # have contract enabled, because it won't be used.
+        # This needs to be executed after contract config is set
+        if self.contract is True:
+            contract_state = ""
+            # We need to sort the columns so that order doesn't matter
+            # columns is a str: ColumnInfo dictionary
+            sorted_columns = sorted(self.columns.values(), key=lambda col: col.name)
+            for column in sorted_columns:
+                contract_state += f"|{column.name}"
+                contract_state += str(column.data_type)
```

Comment:
Makes sense to leave out constraints for the scope of this issue 👍 

We should revisit whether constraints should go in the checksum when we go to implement https://github.com/dbt-labs/dbt-core/issues/7065, where:
* breaking changes to constraints should raise a `ModelContractError` from `same_contract`
* non-breaking changes to constraints (e.g. adding a constraint, removing an unenforced constraint) should return `False` from `same_contract`. 

Determining the nature of a change to a constraint will need additional logic beyond comparing checksums, and adding constraints to the checksum will make understanding changes to column data_type and names a bit trickier. I could see the checksum being a useful to determine, crudely, whether there was a _change at all_, and needing to actually inspect the columns/constraints more closely to determine the type of change - breaking vs non-breaking.

Reply:
What is the benefit of putting these methods on `CompiledNode` versus carrying the columns on `Contract` and overriding `__eq__()` on `Contract` (in place of `same_contract()`)?

Reply:
We are not prepared to move those fields at this time, since it would be a breaking change for packages which use those fields. In addition, not all of the attributes of a column count for breaking changes, so a simple equal doesn't work.